services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,528 @@
+                        .recover(error -> {
+                            if (findQuery.containsKey(RegistrationConstants.FIELD_PAYLOAD_TENANT_ID)) {
+                                return Future.failedFuture(
+                                        new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND,
+                                                String.format("Tenant [%s] not found.",
+                                                        findQuery.getString(
+                                                                RegistrationConstants.FIELD_PAYLOAD_TENANT_ID))));
+                            }
+                            return Future.failedFuture(
+                                    new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND, "Tenant not found"));
+                        }));
+    }
+
+    /**
+     * {@inheritDoc}
+     * @return A future indicating the outcome of the operation.
+     *         The <em>status</em> will be
+     *         <ul>
+     *         <li><em>201 Created</em> if the tenant has been added successfully.</li>
+     *         <li><em>409 Conflict</em> if a tenant with the given
+     *         identifier or certificate authority already exists.</li>
+     *         <li><em>500 Internal Server Error</em> if tenants cannot be accessed.</li>
+     *         </ul>
+     * @throws NullPointerException if the parameters are {@code null}.
+     */
+    @Override
+    public Future<OperationResult<Id>> createTenant(final Optional<String> tenantId, final Tenant tenantObj,
+            final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(tenantObj);
+        Objects.requireNonNull(span);
+
+        final String tenantIdOrGenerated = tenantId.orElse(DeviceRegistryUtils.getUniqueIdentifier());
+
+        return MongoDbDeviceRegistryUtils.isAllowedToModify(config, tenantIdOrGenerated)
+                .compose(ok -> processCreateTenant(tenantIdOrGenerated, tenantObj, span))
+                .recover(error -> Future.succeededFuture(MongoDbDeviceRegistryUtils.mapErrorToResult(error, span)));
+    }
+
+    private Future<OperationResult<Id>> processCreateTenant(final String tenantId, final Tenant tenantObj,
+            final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(tenantObj);
+        Objects.requireNonNull(span);
+
+        final TenantDto newTenantDto = new TenantDto(tenantId, tenantObj,
+                new Versioned<>(tenantObj).getVersion());
+
+        final JsonObject newTenantDtoJson = JsonObject.mapFrom(newTenantDto);
+        final Promise<String> createTenantPromise = Promise.promise();
+        mongoClient.insert(config.getCollectionName(), newTenantDtoJson, createTenantPromise);
+        return createTenantPromise.future()
+                .compose(tenantObjectIdResult -> {
+                    span.log(String.format("successfully created tenant [%s]", tenantId));
```

Comment:
We could also remove logging of the tenant ID using span (in other places too) as the span should already be tagged with the tenant ID.

Reply:
Are you working on it?

Reply:
done